design/kata-api-design.md

```diff
@@ -0,0 +1,143 @@
+# Kata API Design
+To fulfill the [kata design requirements](kata-design-requirements.md), and based on the disscusion on [Virtcontainers API extentions](https://docs.google.com/presentation/d/1dbGrD1h9cpuqAPooiEgtiwWDGCYhVPdatq7owsKHDEQ), the kata runtime library features following APIs:
+-  Sandbox based top API
+-  Storage and network hotplug API
+-  Plugin frameworks for external proprietary Kata runtime extensions
+-  Built-in shim and proxy types and capabilities
+
+## Sandbox Based API
+### Sandbox Management API
+-  CreateSandbox(ID, SandboxConfig)
+
+Create, start a sandbox and return the sandbox structure.
+-  ConnectSandbox(ID)
+
+Connect to an existing sandbox and return the sandbox structure.
+-  ListSandbox()
+
+List all existing standboxes with status.
+
+### Sandbox Operation API
+-  sandbox.Release()
+
+Release the sandbox connection associated with the sandbox structure.
+-  sandbox.Pause()
+
+Pause the sandbox.
+-  sandbox.Resume()
+
+Resume the paused sandbox.
+-  sandbox.Destroy()
+
+Destroy the sandbox and remove all persistent metadata.
+-  sandbox.Stats()
+
+Get the stats of the sandbox.
+-  sandbox.Done()
+
+Return a context handler that caller can monitor for sandbox callbacks such as
+error termination.
+-  sandbox.ListContainer()
+
+List all containers in the sandbox.
+-  sandbox.CreateContainer()
+
+Create new container in the sandbox.
+-  sandbox.DeleteContainer()
+
+Delete a container from the sandbox.
+-  sandbox.StartContainer()
+
+Start a container in the sandbox.
+-  sandbox.StatusContainer()
+
+Get the status of a container in the sandbox.
+-  sandbox.AddProcess()
+
+Run a new process in the sandbox.
+-  sandbox.WaitProcess()
+
+Wait on a process.
+### Sandbox Hotplug API
+-  sandbox.AddStorage()
+
+Add new storage to the sandbox.
+-  sandbox.AddNic()
+
+Add new nic to the sandbox.
+-  sandbox.UpdateRoute()
+
+### Sandbox Relay API
+Update the sandbox route table, e.g. for portmapping support.
+-  sandbox.ContainerTTYResize(containerID, processID, Height, Width)
```

Comment:
This belongs to a specific process, and our caller should already have the `Process` handler in its hand:
```go
func (p *Process) WinsizeProcess(width, height int) error {}
```

Reply:
Again, then it will be a process level API. If we do want to export container and process level API, we need to get a consensus on it and have a way to get container and process structure from a sandbox. I know those exist in virtcontainers but we need to export them as well.

Reply:
Well we need inputs from other folks here, but it's really not a big deal to add functions like `FetchContainer()` or `FetchProcess()`, if we decide to go down this path.